pandas-stubs/core/frame.pyi

```diff
@@ -152,7 +153,7 @@ class _iLocIndexerFrame(_iLocIndexer):
         | tuple[IndexType, int]
         | tuple[IndexType, IndexType]
         | tuple[int, IndexType],
-        value: Scalar | Series | DataFrame | np.ndarray | None,
+        value: Scalar | Series | DataFrame | np.ndarray | NAType | NaTType | None,
```

Comment:
I wouldn't mind having `value: object` as we do not keep track of the type anyway.

Reply:
I'm not sure we should encourage putting arbitrary types inside of a `DataFrame` .  So I want to leave this as is.

Reply:
I'm not sure where pandas-stubs should be on the spectrum of what is documented to the subset of what is also recommended.

In terms of deprecations (basically a recommendation), I'm all for staying closer on the recommendation side, even though the deprecated behavior is still possible. When it comes to what to put in a Series (and especially in a dtype-unaware DataFrame), I feel more comfortable being on the what is documented side. I guess I prefer annotations that have a chance of being ported to pandas (I would probably object to restricting the type of value in pandas).

Reply:
> I'm not sure where pandas-stubs should be on the spectrum of what is documented to the subset of what is also recommended.

I call this the "Goldilocks problem". We don't want the types to be too wide, or too narrow, but to be "just right".  The debate is what is "just right"
> 
> In terms of deprecations (basically a recommendation), I'm all for staying closer on the recommendation side, even though the deprecated behavior is still possible. When it comes to what to put in a Series (and especially in a dtype-unaware DataFrame), I feel more comfortable being on the what is documented side. I guess I prefer annotations that have a chance of being ported to pandas (I would probably object to restricting the type of value in pandas).

I agree that within pandas, what should be in there is what works, which probably means wide types in many places.  We'll have to see how this evolves over time.  My concern with having wide types is that it doesn't help users catch certain bugs via type checking.